File: app/src/main/java/com/michel/deutschmacht/progress/CalendarView30.java
package com.michel.deutschmacht.progress;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;

import com.michel.deutschmacht.data.Store;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

/** Last-30-days activity grid: green square = active day, today outlined. */
public class CalendarView30 extends View {

    private Set<String> active = new HashSet<>();
    private final Paint fill = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint empty = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint stroke = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint text = new Paint(Paint.ANTI_ALIAS_FLAG);

    public CalendarView30(Context c) { this(c, null); }
    public CalendarView30(Context c, AttributeSet a) { super(c, a); init(c); }

    private void init(Context c) {
        empty.setColor(0x22000000);
        fill.setColor(0xFF2E7D5B);
        stroke.setColor(0xFFF2A94C);
        stroke.setStyle(Paint.Style.STROKE);
        stroke.setStrokeWidth(4f);
        text.setColor(0xFF1B4D3E);
        text.setTextSize(30f);
        text.setTextAlign(Paint.Align.CENTER);
    }

    public void refresh(Store store) {
        active = new HashSet<>();
        for (String d : store.activeDays()) active.add(d);
        invalidate();
    }

    @Override
    protected void onMeasure(int w, int h) {
        int cols = 10, rows = 3;
        int side = (Math.max(MeasureSpec.getSize(w), 300) - (cols - 1) * 8) / cols;
        setMeasuredDimension(getDefaultSize(getSuggestedMinimumWidth(), w), rows * side + (rows - 1) * 8);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int cols = 10;
        float side = (getWidth() - (cols - 1) * 8f) / cols;
        SimpleDateFormat f = new SimpleDateFormat("yyyy-MM-dd", Locale.US);
        Calendar cal = Calendar.getInstance();
        cal.add(Calendar.DAY_OF_MONTH, -29);
        for (int i = 0; i < 30; i++) {
            int col = i % cols, row = i / cols;
            float x = col * (side + 8);
            float y = row * (side + 8);
            String d = f.format(cal.getTime());
            canvas.drawRoundRect(new RectF(x, y, x + side, y + side), 8, 8,
                    active.contains(d) ? fill : empty);
            if (i == 29) canvas.drawRoundRect(new RectF(x, y, x + side, y + side), 8, 8, stroke);
            cal.add(Calendar.DAY_OF_MONTH, 1);
        }
    }
}

File: app/src/main/java/com/michel/deutschmacht/progress/CalendarView30.kt
package com.michel.deutschmacht.progress

import android.content.Context
import android.graphics.Canvas
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import com.michel.deutschmacht.R
import com.michel.deutschmacht.data.Store
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

/** Last-30-days activity grid: filled square = active day, today outlined in orange. */
class CalendarView30 @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : View(context, attrs) {

    private val active = HashSet<String>()

    private val fill = android.graphics.Paint(android.graphics.Paint.ANTI_ALIAS_FLAG).apply {
        color = context.getColor(R.color.fill_active)
    }
    private val empty = android.graphics.Paint(android.graphics.Paint.ANTI_ALIAS_FLAG).apply {
        color = context.getColor(R.color.fill_empty)
    }
    private val stroke = android.graphics.Paint(android.graphics.Paint.ANTI_ALIAS_FLAG).apply {
        color = context.getColor(R.color.today_stroke)
        style = android.graphics.Paint.Style.STROKE
        strokeWidth = 4f
    }

    private val cols = 10
    private val rows = 3
    private val gap = 8f

    fun refresh(store: Store) {
        active.clear()
        active.addAll(store.activeDays().toList())
        invalidate()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val side = (maxOf(MeasureSpec.getSize(widthMeasureSpec), 300) - (cols - 1) * gap.toInt()) / cols
        setMeasuredDimension(
            getDefaultSize(suggestedMinimumWidth, widthMeasureSpec),
            rows * side + (rows - 1) * gap.toInt()
        )
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val side = (width - (cols - 1) * gap) / cols
        val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        val cal = Calendar.getInstance()
        cal.add(Calendar.DAY_OF_MONTH, -29)
        for (i in 0 until 30) {
            val col = i % cols
            val row = i / cols
            val x = col * (side + gap)
            val y = row * (side + gap)
            val rect = RectF(x, y, x + side, y + side)
            canvas.drawRoundRect(rect, 8f, 8f, if (active.contains(fmt.format(cal.time))) fill else empty)
            if (i == 29) canvas.drawRoundRect(rect, 8f, 8f, stroke)
            cal.add(Calendar.DAY_OF_MONTH, 1)
        }
    }
}
